state/upgrades.go

```diff
@@ -150,3 +151,65 @@ func RemoveOrphanedSecretPermissions(pool *StatePool) error {
 		return nil
 	}))
 }
+
+// MigrateApplicationOpenedPortsToUnitScope copies the opened ports for an application for all its units.
+func MigrateApplicationOpenedPortsToUnitScope(pool *StatePool) error {
+	return errors.Trace(runForAllModelStates(pool, func(st *State) error {
+		apps, err := st.AllApplications()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		var appNames []string
+		for _, app := range apps {
+			appNames = append(appNames, app.Name())
+		}
+		openedPorts, closer := st.db().GetCollection(openedPortsC)
+		defer closer()
+		docs := []applicationPortRangesDoc{}
+		if err = openedPorts.Find(bson.D{{"application-name", bson.D{{"$in", appNames}}}}).All(&docs); err != nil {
+			return errors.Trace(err)
+		}
+
+		var ops []txn.Op
+		for _, doc := range docs {
+			if len(doc.PortRanges) == 0 {
+				continue
+			}
+
+			app, err := st.Application(doc.ApplicationName)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			units, err := app.AllUnits()
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if len(units) == 0 {
+				continue
+			}
+
+			if doc.UnitRanges == nil {
+				doc.UnitRanges = make(map[string]network.GroupedPortRanges)
+			}
+			for _, unit := range units {
+				doc.UnitRanges[unit.Name()] = doc.PortRanges.Clone()
+			}
+			doc.PortRanges = nil
+
+			ops = append(ops, txn.Op{
+				C:  openedPortsC,
+				Id: doc.DocID,
+				Update: bson.D{
+					{Name: "$set", Value: bson.D{
+						{Name: "port-ranges", Value: doc.PortRanges},
```

Comment:
Do we need "port-ranges"? We are just treating the ports as being the same for all units right?
So we copy the ports to all units and leave any application ports empty till next cycle.

Reply:
doc.PortRanges has already set to nil.